---
title: "eQTL_comparison"
author: "Francis Leblanc"
date: "21/01/2023"
output:   
  html_document:
    toc: yes
    toc_float: yes
    code_folding: hide 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
set.seed(2023)
```

# lib 

```{r, echo = FALSE, warning = FALSE, results = 'hide', message = FALSE}
library(parallel)
library(Biobase)
library(snpStats)
library(ggplot2)
library(psych)
library(Hmisc)
library(rtracklayer)
library(dplyr)
library(limma)
```

# Compare eQTLs

## Make table for Significant eQTLs (Table S2)

```{r}
# plink --bfile data/Harbin/plink/eQTL_250kb --freq --out data/Harbin/plink/eQTL_250kb.freq
# plink --bfile data/CTSN/plink/eQTL_250kb --freq --out data/CTSN/plink/eQTL_250kb.freq
```


```{r}
freq.harbin <- read.table("data/Harbin/plink/eQTL_250kb.freq.frq", header = T)
freq.ctsn <- read.table("data/CTSN/plink/eQTL_250kb.freq.frq", header = T)

h.cis <- readRDS("results/Harbin_cis_eQTLs.sex.7PCs.RDS")
c.cis <- readRDS("results/CTSN_cis_eQTLs.sex.7PCs.RDS")

# Merge the Harbin and CTSN cis-eQTLs
eqtls <- merge(h.cis, 
               c.cis,
               by = c('snps',
                      'gene',
                      'GENEID',
                      'gene_name',
                      'seqnames',
                      'start',
                      'gene_type',
                      'snps_gene',
                      'rsid',
                      'Meta_P'), 
               all = T,
               suffixes = c("_Harbin","_CTSN"))

freq <- merge(freq.harbin[, -c(1,6)], 
              freq.ctsn[, -c(1,6)], 
              by = "SNP", 
              all = T,
              suffixes = c("_harbin","_CTSN"))

eqtls <- merge(eqtls, 
               freq, 
               by.x = "snps",
               by.y = "SNP", 
               all.x = T)

rio::export(eqtls, "results/eQTL.tab.csv") # Table S2 
```

## Comparisons of effects CTSN vs Harbin (Figure 1C)

```{r}
# Identify SNPs with flipped alleles
to.flip <- which(eqtls$A1_harbin != eqtls$A1_CTSN)
eqtls.f <- eqtls
eqtls.f$beta_Harbin[to.flip] <- -1 * eqtls.f$beta_Harbin[to.flip]

# genes to label on plot (top 20 each dataset)
lab.gene <- slice_min(c.cis, FDR, n = 20) %>%
  pull(snps_gene)

lab.gene <- slice_min(h.cis, FDR, n = 20) %>%
  pull(snps_gene) %>% 
  c(., lab.gene)

# Show significance in each dataset
eqtls.f$signif <- "Not significant"
eqtls.f$signif <- ifelse(eqtls.f$FDR_Harbin < 0.05, "Harbin", eqtls.f$signif)
eqtls.f$signif <- ifelse(eqtls.f$FDR_CTSN < 0.05, "CTSN", eqtls.f$signif)
eqtls.f$signif <- ifelse(eqtls.f$FDR_Harbin < 0.05 & eqtls.f$FDR_CTSN < 0.05, 
                         "Both", eqtls.f$signif)

# compute standard error & add genes to label
eqtls.f <- eqtls.f %>%
  mutate(eQTL = ifelse(snps_gene %in% lab.gene, snps_gene, NA),
         eQTL.gene = strsplit2(eQTL, "_")[, 2],
         eQTL.gene = ifelse(eQTL.gene == "", NA, eQTL.gene),
         beta_se_Harbin = beta_Harbin / statistic_Harbin,
         beta_se_CTSN = beta_CTSN / statistic_CTSN,
         beta_se_Harbin = ifelse(signif == "Not significant", 
                                 NA, 
                                 beta_se_Harbin),
         beta_se_CTSN = ifelse(signif == "Not significant",
                               NA, 
                               beta_se_CTSN), 
         signif = factor(signif, levels = c('Not significant', 
                                            'Both', 
                                            'CTSN', 
                                            'Harbin'))) %>% 
  arrange(signif)

eqtls.f %>% 
  dplyr::filter(signif != "Not significant") %>% 
  ggplot(aes(beta_Harbin, beta_CTSN, label = eQTL.gene, color = signif)) +
  geom_abline(slope = 1, linetype = "dashed") +
  geom_point(size = 1) +
  geom_errorbarh(aes(xmax = beta_Harbin + beta_se_Harbin, 
                     xmin = beta_Harbin - beta_se_Harbin),
                 alpha = 0.4) +
  geom_errorbar(aes(ymax = beta_CTSN + beta_se_CTSN,
                    ymin = beta_CTSN - beta_se_CTSN),
                alpha = 0.4) +
  cowplot::theme_cowplot() +
  ggrepel::geom_text_repel(size = 3, 
                           max.overlaps = 50,
                           min.segment.length = 0.01) +
  ggtitle(paste0(nrow(eqtls.f), " tested cis-eQTL"), 
          subtitle = "Significant in both = 11\nHarbin only = 6\nCTSN only = 16")+
  theme(legend.position = c(0.7, 0.2)) + 
  ylab("beta CTSN") +
  xlab("beta Harbin") +
  scale_color_manual(values = c("darkred", "darkgreen", "blue"))

ggsave("figs/eQTL_comparison/compared_eQTLs.significant.betas.flipped.RNA.PCs.png",
       width = 6,
       height = 6)
```

## Pull GTEx P-values

```{r}
RAA.gtex <- read.table(
  "data/external/Heart_Atrial_Appendage.v8.signif_variant_gene_pairs.txt",
  header = T
  )
RAA.gtex$gene_id <- strsplit2(RAA.gtex$gene_id, "[.]")[, 1]
snp.ID <- strsplit2(RAA.gtex$variant_id, "_")[, 1:2]
RAA.gtex$snp.ID <- paste0(snp.ID[, 1], ":", snp.ID[, 2])
RAA.gtex$snp_gene <- paste0(RAA.gtex$snp.ID, "_", RAA.gtex$gene_id)

snp.ID <- strsplit2(eqtls.f$snps, ":")[, 1:2]
eqtls.f$snp.ID <- paste0(snp.ID[, 1], ":", snp.ID[, 2])
eqtls.f$snp_gene <- paste0(eqtls.f$snp.ID, "_", eqtls.f$GENEID)

RAA.gtex.f <- RAA.gtex[match(eqtls.f$snp_gene, RAA.gtex$snp_gene),
                       c("variant_id",
                         "tss_distance",
                         "maf",
                         "pval_nominal",
                         "pval_nominal_threshold",
                         "min_pval_nominal")]

colnames(RAA.gtex.f) <- paste0("RAA.gtex_", colnames(RAA.gtex.f))
eqtls.f.gtex <- cbind(eqtls.f, RAA.gtex.f)
eqtls.f.gtex$snp_hg38.pos <- as.numeric(strsplit2(eqtls.f.gtex$snps, ":")[, 2])

rio::export(eqtls.f.gtex, "results/flipped.eQTLs.table.GWAS.GTEx.csv")
```

## GTEx vs CTSN (Figure 1D)

```{r}
# merge datasets
RAA.gtex.f <- RAA.gtex[match(eqtls.f$snp_gene, RAA.gtex$snp_gene),
                       c("variant_id",
                         "tss_distance",
                         "maf",
                         "pval_nominal",
                         "pval_nominal_threshold",
                         "min_pval_nominal",
                         "slope",
                         "slope_se")]

colnames(RAA.gtex.f) <- paste0("RAA.gtex_", colnames(RAA.gtex.f))
RAA.gtex.f$RAA.gtex_ref <- strsplit2(RAA.gtex.f$RAA.gtex_variant_id, "_")[, 3]
eqtls.f.gtex <- cbind(eqtls.f, RAA.gtex.f)

# flip slope if ref allele doesn't match and fill NA values
eqtls.f.gtex <- eqtls.f.gtex %>% 
  mutate(slope_fliped = ifelse(RAA.gtex_ref != A1_CTSN,
                               RAA.gtex_slope * -1, 
                               RAA.gtex_slope),
         slope_fliped = ifelse(is.na(RAA.gtex_slope),
                               0,
                               slope_fliped),
         RAA.gtex_pval_nominal = ifelse(is.na(RAA.gtex_pval_nominal),
                               1,
                               RAA.gtex_pval_nominal),
         RAA.gtex_pval_nominal_threshold = ifelse(
           is.na(RAA.gtex_pval_nominal_threshold),
           max(RAA.gtex_pval_nominal_threshold, na.rm = T),
                               RAA.gtex_pval_nominal_threshold
           ))

# set significance labels
eqtls.f.gtex$signif <- "Not significant"
eqtls.f.gtex <- eqtls.f.gtex %>% 
  mutate(FDR_CTSN = ifelse(is.na(FDR_CTSN), 1, FDR_CTSN),
         signif = ifelse(RAA.gtex_pval_nominal < RAA.gtex_pval_nominal_threshold,
                              "GTEx", 
                              signif),
         signif = ifelse(FDR_CTSN < 0.05, "CTSN", signif),
         signif = ifelse(RAA.gtex_pval_nominal < RAA.gtex_pval_nominal_threshold & 
                           FDR_CTSN < 0.05, "Both", signif))


# label top eQTL genes  
lab.gene <- slice_min(eqtls.f.gtex, pvalue_CTSN, n = 20) %>%
  pull(snps_gene)
lab.gene <- slice_min(eqtls.f.gtex, RAA.gtex_pval_nominal, n = 20) %>%
  pull(snps_gene) %>%
  c(., lab.gene)

eqtls.f.gtex$eQTL.gene <- ifelse(eqtls.f.gtex$snps_gene %in% lab.gene,
                                 eqtls.f.gtex$gene_name, 
                                 NA)
eqtls.f.gtex$beta_se_CTSN <- eqtls.f.gtex$beta_CTSN / eqtls.f.gtex$statistic_CTSN

# betas
eqtls.f.gtex %>% 
  dplyr::filter(!(slope_fliped == 0 & FDR_CTSN > 0.05)) %>% 
  ggplot(aes(slope_fliped, beta_CTSN, label = eQTL.gene, color = signif)) +
  geom_errorbarh(aes(xmax = slope_fliped + RAA.gtex_slope_se, 
                     xmin = slope_fliped - RAA.gtex_slope_se), alpha = 0.4) +
  geom_errorbar(aes(ymax = beta_CTSN + beta_se_CTSN, 
                    ymin = beta_CTSN - beta_se_CTSN), alpha = 0.4) +
  geom_abline(slope = 1, linetype = "dashed") +
  geom_point(size = 1) +
  cowplot::theme_cowplot() +
  ggrepel::geom_text_repel(size = 3,
                           max.overlaps = 50,
                           min.segment.length = 0.01) +
  ggtitle(paste0(nrow(c.cis)," tested cis-eQTL in CTSN"), 
          subtitle = "Significant in both = 19\nGTEx only = 97\nCTSN only = 9") + 
  theme(legend.position = c(0.7,0.2)) + 
  ylab("beta CTSN") + 
  xlab("beta GTEx") + 
  scale_color_manual(values = c("darkred", "darkgreen", "black"))

ggsave("figs/eQTL_comparison/compared_eQTLs.betas.flipped.GTEx_CTSN_slope.png",
       width = 6,
       height = 6)
```

## GTEx vs Harbin (Figure 1E)

```{r}
# set significance labels
eqtls.f.gtex$signif <- "Not significant"
eqtls.f.gtex <- eqtls.f.gtex %>% 
  mutate(FDR_Harbin = ifelse(is.na(FDR_Harbin), 1, FDR_Harbin),
         signif = ifelse(RAA.gtex_pval_nominal < RAA.gtex_pval_nominal_threshold,
                         "GTEx", 
                         signif),
         signif = ifelse(FDR_Harbin < 0.05, "Harbin", signif),
         signif = ifelse(RAA.gtex_pval_nominal < RAA.gtex_pval_nominal_threshold &
                           FDR_Harbin < 0.05, "Both", signif))

# label top eQTL genes  
lab.gene <- slice_min(eqtls.f.gtex, pvalue_Harbin, n = 20) %>%
  pull(snps_gene)
lab.gene <- slice_min(eqtls.f.gtex, RAA.gtex_pval_nominal, n = 20) %>%
  pull(snps_gene) %>%
  c(., lab.gene)

eqtls.f.gtex$eQTL.gene <- ifelse(eqtls.f.gtex$snps_gene %in% lab.gene,
                                 eqtls.f.gtex$gene_name, 
                                 NA)

# betas
eqtls.f.gtex$beta_se_Harbin <- eqtls.f.gtex$beta_Harbin / eqtls.f.gtex$statistic_Harbin
eqtls.f.gtex %>% 
  dplyr::filter(!(slope_fliped == 0 & FDR_Harbin > 0.05)) %>% 
  ggplot(aes(slope_fliped, beta_Harbin, label = eQTL.gene, color = signif)) +
  geom_errorbarh(aes(xmax = slope_fliped + RAA.gtex_slope_se, 
                     xmin = slope_fliped - RAA.gtex_slope_se), alpha = 0.4) +
  geom_errorbar(aes(ymax = beta_Harbin + beta_se_Harbin,
                    ymin = beta_Harbin - beta_se_Harbin), alpha = 0.4) +
  geom_abline(slope = 1, linetype = "dashed") +
  geom_point(size = 1) +
  cowplot::theme_cowplot() +
  ggrepel::geom_text_repel(size = 3, max.overlaps = 50, min.segment.length = 0.01) +
  ggtitle(paste0(nrow(h.cis)," tested cis-eQTL in Harbin"), 
          subtitle = "Significant in both = 14\nGTEx only = 102\nHarbin only = 3")+
  theme(legend.position = c(0.7, 0.2)) + 
  ylab("beta Harbin") + 
  xlab("beta GTEx") + 
  scale_color_manual(values = c("darkred", "black", "blue"))

ggsave("figs/eQTL_comparison/compared_eQTLs.betas.flipped.GTEx_Harbin_slope.png",
       width = 6,
       height = 6) 
```

# Compare interactions 

```{r}
inter.h <- readRDS("results/Harbin_cis_interactions.7PCs.RDS") 
inter.c <- readRDS("results/CTSN_cis_interactions.7PCs.RDS") 
```

## QQ plot (Figure S3) 

```{r}
plot_qq <- function(cis_eqtls, cohort) {
  cis_eqtls <- cis_eqtls[order(cis_eqtls$`genotype:condition`),]
  df.ci<- data.frame(
    expected = -log10(ppoints(nrow(cis_eqtls))),
    clower   = -log10(qbeta(p = (1 - 0.95) / 2, 
                            shape1 = 1:nrow(cis_eqtls),
                            shape2 = nrow(cis_eqtls):1)),
    cupper   = -log10(qbeta(p = (1 + 0.95) / 2, 
                            shape1 = 1:nrow(cis_eqtls),
                            shape2 = nrow(cis_eqtls):1)),
    pvalue = -log10(cis_eqtls$`genotype:condition`),
    gene = cis_eqtls$gene_name
  )
  
  df.ci$gene[20:nrow(df.ci)] <- NA
  
  ggplot(df.ci) +
    geom_point(aes(x = expected, y = pvalue), alpha = 0.5, size = 1) +
    ggrepel::geom_text_repel(aes(x = expected, y = pvalue, label = gene),
                             size = 2,
                             max.overlaps = 30) +
    geom_abline(intercept = 0, slope = 1, alpha = 0.5) +
    geom_line(data = df.ci, aes(expected, cupper), linetype = 2) +
    geom_line(data = df.ci, aes(expected, clower), linetype = 2) +
    theme_minimal() +
    ylab(substitute(paste("Observed -log10(", italic("P"), ")"))) +
    xlab(substitute(paste("Expected -log10(", italic("P"), ")"))) +
    ggtitle(paste0(cohort, " SNP:AF interactions"))
}
plot_qq(inter.c, "CTSN")
ggsave("figs/eQTL_comparison/CTSN_cis.eQTL_interactions_QQ-plot.png",
       width = 6,
       height = 4)

plot_qq(inter.h, "Harbin")
ggsave("figs/eQTL_comparison/Harbin_cis.eQTL_interactions_QQ-plot.png",
       width = 6, 
       height = 4)
```

# save loci with significant eQTL in CTSN or Harbin for coloc

```{r}
cis_eqtls.h <- readRDS("data/Harbin/MatrixEQTL/250kb_eQTLs_7PCs.RDS")$cis$eqtls
cis_eqtls.c <- readRDS("data/CTSN/MatrixEQTL/250kb_eQTLs_7PCs.RDS")$cis$eqtls
eQTL <- rio::import("results/flipped.eQTLs.table.GWAS.GTEx.csv")
eQTL <- arrange(eQTL, FDR_CTSN) %>% 
  distinct(snps, .keep_all = T)

gr <- GRanges(seqnames = eQTL$seqnames,
              ranges = IRanges(start = eQTL$snp_hg38.pos - 250000,
                               end = eQTL$snp_hg38.pos + 250000),
              rsid = eQTL$rsid)

# find significant genes by locus
sign.genes.locus <- function(eQTL.res, loci.gr) {
  snp_id <- strsplit2(eQTL.res$snps, ":")[, c(1, 2)] %>%
    as.data.frame()
  
  gr.eqtl <- GRanges(seqnames = snp_id$V1,
                     ranges = IRanges(start = as.numeric(snp_id$V2),
                                      end = as.numeric(snp_id$V2)))
  
  gr.eqtl$FDR <- eQTL.res$FDR
  gr.eqtl$gene <- eQTL.res$gene
  
  # return unique significant genes in the locus for each eQTL
  l.genes <- lapply(1:length(loci.gr), function(x) {
    gr.s <- subsetByOverlaps(gr.eqtl, loci.gr[x])
    unique(gr.s[gr.s$FDR < 0.05]$gene)
  })
  names(l.genes) <- loci.gr$rsid
  return(l.genes)
}

harbin_hits <- sign.genes.locus(cis_eqtls.h, gr)
ctsn_hits <- sign.genes.locus(cis_eqtls.c, gr)

saveRDS(harbin_hits, "data/Harbin/harbin_hits.list.rds")
saveRDS(ctsn_hits, "data/CTSN/ctsn_hits.list.rds")

l.union <- lapply(names(harbin_hits), function(x) {
  union(harbin_hits[[x]], ctsn_hits[[x]])
})
names(l.union) <- gr$rsid
saveRDS(l.union, "data/eQTL.union_hits.list.rds")
```
